File: Java Barcode Scanner ME/app/src/main/java/de/dimedis/mobileentry/adapter/ItemAdapter.java
package de.dimedis.mobileentry.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import de.dimedis.mobileentry.R;

public class ItemAdapter extends RecyclerView.Adapter<ItemAdapter.ViewHolder> {
    private List<Data> mDataset;

  public static class Data{
    public String name;
      public String code;

      public Data(String code, String name) {
          this.name = name;
          this.code = code;
      }
  }


    public static class ViewHolder extends RecyclerView.ViewHolder {
        public TextView mTextView;
        public ImageView mImageView;

        public ViewHolder(View v) {
            super(v);
            mTextView = (TextView)v.findViewById(R.id.label);
            mImageView = (ImageView)v.findViewById(R.id.icon);
        }
    }


    public ItemAdapter(List<Data> mDataset) {
        this.mDataset = mDataset;
    }

    // Create new views (invoked by the layout manager)
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent,
                                                   int viewType) {
        // create a new view
        View v  = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_language, parent, false);

        ViewHolder vh = new ViewHolder(v);
        return vh;
    }


public static interface  OnItemClick{
        void onClick(ViewHolder holder,Data dat, int position);
    }
    OnItemClick mOnItemClick;
    public void   setOnItemClick(OnItemClick onItemClick){
         this.mOnItemClick=onItemClick;
     }

    @Override
    public void onBindViewHolder(final ViewHolder holder, final int position) { // TODO Error
        holder.mTextView.setText(mDataset.get(position).name);
        if(mOnItemClick!=null) {
            holder.itemView.setOnClickListener(v -> mOnItemClick.onClick(holder, getItem(position), position)
            );
        };

    }

    @Override
    public int getItemCount() {
        return mDataset.size();
    }


    public Data getItem(int position) {
              return mDataset.get(position);

    }
}

File: Kotlin Migration/NTMobileEntry/app/src/main/java/de/dimedis/mobileentry/adapter/ItemAdapter.kt
package de.dimedis.mobileentry.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import de.dimedis.mobileentry.R

class ItemAdapter(private val mDataset: List<Data>) : RecyclerView.Adapter<ItemAdapter.ViewHolder>() {
    class Data(var code: String, var name: String)
    class ViewHolder(v: View) : RecyclerView.ViewHolder(v) {
        var mTextView: TextView
        var mImageView: ImageView

        init {
            mTextView = v.findViewById<View>(R.id.label) as TextView
            mImageView = v.findViewById<View>(R.id.icon) as ImageView
        }
    }

    // Create new views (invoked by the layout manager)
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        // create a new view
        val v = LayoutInflater.from(parent.context).inflate(R.layout.item_language, parent, false)
        return ViewHolder(v)
    }

    interface OnItemClick {
        fun onClick(holder: ViewHolder?, dat: Data?, position: Int)
    }

    var mOnItemClick: OnItemClick? = null
    fun setOnItemClick(onItemClick: OnItemClick?) {
        mOnItemClick = onItemClick
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.mTextView.text = mDataset[position].name
        if (mOnItemClick != null) {
            holder.itemView.setOnClickListener {
                mOnItemClick!!.onClick(holder, getItem(position), position)
            }
        }
    }

    override fun getItemCount(): Int {
        return mDataset.size
    }

    fun getItem(position: Int): Data {
        return mDataset[position]
    }
}